In [8]:
# import dependencies and load data
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score, classification_report
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import accuracy_score
import numpy as np

In [9]:
# Load and preprocess the training data
url = "https://raw.githubusercontent.com/WHPAN0108/BHT-DataScience-S23/main/classification/data/Assigment/aug_train.csv"
train_data = pd.read_csv(url)
train_data['experience'] = train_data['experience'].replace(['>20', '<1'], [21, 1])
train_data['last_new_job'] = train_data['last_new_job'].replace(['>4', 'never'], [5, 0])
categorical_columns = ['gender', 'relevent_experience', 'enrolled_university', 'education_level', 'major_discipline', 'company_type']
for column in categorical_columns:
    train_data[column].fillna(train_data[column].mode()[0], inplace=True)
numerical_columns = ['city_development_index', 'training_hours']
for column in numerical_columns:
    train_data[column].fillna(train_data[column].median(), inplace=True)
X_train = train_data.drop('target', axis=1)
y_train = train_data['target']
X_train = pd.get_dummies(X_train)
X_train_split, X_val_split, y_train_split, y_val_split = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# Train the Random Forest classifier
rf_classifier = RandomForestClassifier()
rf_classifier.fit(X_train_split, y_train_split)

# Predict the target variable for the validation set
y_val_pred = rf_classifier.predict(X_val_split)

a = rf_classifier.predict(X_train)
print("test:", len(a), a[0:20])

# Evaluate the accuracy of the model
accuracy = accuracy_score(y_val_split, y_val_pred)
print("- - -")
print("Validation Accuracy:", accuracy)
print("- - -")


test: 2100 [0 0 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 1 0]
- - -
Validation Accuracy: 0.7261904761904762
- - -


In [10]:
url = "https://raw.githubusercontent.com/WHPAN0108/BHT-DataScience-S23/main/classification/data/Assigment/aug_train.csv"
train_data = pd.read_csv(url)

# True labels from the test set
y_true = train_data['target']

# Compute the confusion matrix
conf_matrix = confusion_matrix(y_true, a)

# Calculate the accuracy
accuracy = accuracy_score(y_true, a)

# Calculate the precision
precision = precision_score(y_true, a)

# Calculate the recall
recall = recall_score(y_true, a)

# Calculate the F1-score
f1 = f1_score(y_true, a)

# Print the results
print("Confusion Matrix:\n", conf_matrix)
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Confusion Matrix:
 [[1533   32]
 [  83  452]]
Accuracy: 0.9452380952380952
Precision: 0.9338842975206612
Recall: 0.8448598130841122
F1-score: 0.8871442590775269


In [11]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier

# Load the test dataset
url = "https://raw.githubusercontent.com/WHPAN0108/BHT-DataScience-S23/main/classification/data/Assigment/aug_test.csv"
test_data = pd.read_csv(url)
test_data['experience'] = test_data['experience'].replace(['>20', '<1'], [21, 1])
test_data['last_new_job'] = test_data['last_new_job'].replace(['>4', 'never'], [5, 0])
categorical_columns = ['gender', 'relevent_experience', 'enrolled_university', 'education_level', 'major_discipline', 'company_type']
for column in categorical_columns:
    test_data[column].fillna(test_data[column].mode()[0], inplace=True)
numerical_columns = ['city_development_index', 'training_hours']
for column in numerical_columns:
    test_data[column].fillna(test_data[column].median(), inplace=True)

# Reorder columns to match the training dataset
test_data = test_data.reindex(columns=X_val_split.columns, fill_value=0)

# Predict the target variable for the new dataset
y_pred = rf_classifier.predict(test_data)

# Print the predicted values
print("Predicted targets:", len(y_pred), y_pred)

Predicted targets: 100 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]


In [12]:
url = "https://raw.githubusercontent.com/WHPAN0108/BHT-DataScience-S23/main/classification/data/Assigment/aug_test.csv"
test_data = pd.read_csv(url)

# True labels from the test set
y_true = test_data['target']

# Compute the confusion matrix
conf_matrix = confusion_matrix(y_true, y_pred)

# Calculate the accuracy
accuracy = accuracy_score(y_true, y_pred)

# Calculate the precision
precision = precision_score(y_true, y_pred)

# Calculate the recall
recall = recall_score(y_true, y_pred)

# Calculate the F1-score
f1 = f1_score(y_true, y_pred)

# Print the results
print("Confusion Matrix:\n", conf_matrix)
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Confusion Matrix:
 [[78  0]
 [21  1]]
Accuracy: 0.79
Precision: 1.0
Recall: 0.045454545454545456
F1-score: 0.08695652173913045


To compare the results between the training and test sets, we can examine the performance metrics such as accuracy, precision, recall, and F1-score for both sets. Let's compare the results obtained from the training set and the test set:

Training Set:

    Confusion Matrix:
    [[1533 32]
    [ 83 452]]
    Accuracy: 0.9452380952380952
    Precision: 0.9338842975206612
    Recall: 0.8448598130841122
    F1-score: 0.8871442590775269

Test Set:

    Confusion Matrix:
    [[78 0]
    [21 1]]
    Accuracy: 0.79
    Precision: 1.0
    Recall: 0.045454545454545456
    F1-score: 0.08695652173913045

Based on the results, we can observe the following differences between the training and test sets:

    Accuracy: The accuracy on the training set is 0.9452, while the accuracy on the test set is 0.79. The model performs significantly better on the training set compared to the test set, indicating a potential overfitting issue.

    Precision: The precision on the training set is 0.9339, indicating a high percentage of correctly predicted positive cases. However, on the test set, the precision is 1.0, which means all predicted positive cases are correct. This could be due to the small number of positive cases in the test set (only 1), resulting in an inflated precision value.

    Recall: The recall on the training set is 0.8449, suggesting that the model can identify a good proportion of positive cases. However, on the test set, the recall is only 0.0455, indicating that the model struggles to identify positive cases in the test set.

    F1-score: The F1-score on the training set is 0.8871, which is a harmonic mean of precision and recall. On the test set, the F1-score is 0.0870, indicating a low balance between precision and recall.

In summary, the model performs well on the training set with high accuracy, precision, recall, and F1-score. However, when applied to the test set, the performance drops significantly, indicating a potential issue with generalization. It is important to address this overfitting and evaluate the model's performance on unseen data to ensure its reliability and effectiveness.